src/desktop/space.rs

```diff
@@ -0,0 +1,419 @@
+            render_scale: scale,
+            ..Default::default()
+        };
+        if !self.outputs.contains(output) {
+            self.outputs.push(output.clone());
+        }
+    }
+
+    pub fn outputs(&self) -> impl Iterator<Item = &Output> {
+        self.outputs.iter()
+    }
+
+    pub fn unmap_output(&mut self, output: &Output) {
+        if let Some(map) = output.user_data().get::<OutputUserdata>() {
+            map.borrow_mut().remove(&self.id);
+        }
+        self.outputs.retain(|o| o != output);
+    }
+
+    pub fn output_geometry(&self, o: &Output) -> Option<Rectangle<i32, Logical>> {
+        if !self.outputs.contains(o) {
+            return None;
+        }
+
+        let state = output_state(self.id, o);
+        o.current_mode().map(|mode| {
+            Rectangle::from_loc_and_size(
+                state.location,
+                mode.size.to_f64().to_logical(state.render_scale).to_i32_round(),
+            )
+        })
+    }
+
+    pub fn output_scale(&self, o: &Output) -> Option<f64> {
+        if !self.outputs.contains(o) {
+            return None;
+        }
+
+        let state = output_state(self.id, o);
+        Some(state.render_scale)
+    }
+
+    pub fn output_for_window(&self, w: &Window) -> Option<Output> {
+        if !self.windows.contains(w) {
+            return None;
+        }
+
+        let w_geo = self.window_geometry(w).unwrap();
+        for o in &self.outputs {
+            let o_geo = self.output_geometry(o).unwrap();
+            if w_geo.overlaps(o_geo) {
+                return Some(o.clone());
+            }
+        }
+
+        // TODO primary output
+        self.outputs.get(0).cloned()
+    }
+
+    pub fn cleanup(&mut self) {
+        self.windows.retain(|w| w.toplevel().alive());
+    }
+
+    pub fn render_output<R>(
+        &mut self,
+        renderer: &mut R,
+        output: &Output,
+        age: usize,
+        clear_color: [f32; 4],
+    ) -> Result<bool, RenderError<R>>
+    where
+        R: Renderer + ImportAll,
+        R::TextureId: 'static,
+    {
+        let mut state = output_state(self.id, output);
+        let output_size = output
+            .current_mode()
+            .ok_or(RenderError::OutputNoMode)?
+            .size
+            .to_f64()
+            .to_logical(state.render_scale)
+            .to_i32_round();
+        let output_geo = Rectangle::from_loc_and_size(state.location, output_size);
+
+        // This will hold all the damage we need for this rendering step
+        let mut damage = Vec::<Rectangle<i32, Logical>>::new();
+        // First add damage for windows gone
+        for old_window in state
+            .last_state
+            .iter()
+            .filter_map(|(id, w)| {
+                if !self.windows.iter().any(|w| w.0.id == *id) {
+                    Some(*w)
+                } else {
+                    None
+                }
+            })
+            .collect::<Vec<Rectangle<i32, Logical>>>()
```

Comment:
What about moving the `debug!()` into the `filter_map()` closure, and collecting the result directly into `damage`, rather than manually iterating & pushing every element?

Reply:
good catch!